syntax="proto3";
package spkdfs;
option cc_generic_services = true;

message Request {
};

message Response {
  bool success = 1;
  string fail_info = 2;
  string redirect = 3;
};

message CommonResponse {
  Response common = 1;
};

message NNLsRequest {
  string path = 1;
}
message NNLsResponse {
  Response common = 1;
  repeated string data = 2;
};

message NNMkdirRequest {
  string path = 1;
}

message NNGetRequest {
  string path = 1;
}
message NNGetResponse {
  Response common = 1;
  repeated uint32 blkids = 2;
};

message NNPutRequest {
  string path = 1;
  repeated uint32 blkids = 2;
  uint64 size = 3;
}
message NNPutResponse {
  Response common = 1;
  repeated string nodes = 2;
};


message NNPutOKRequest {
  string blkid = 1;
};

message NNGetMasterResponse {
  Response common = 1;
  string node = 2;
}
service NamenodeService {
  rpc ls(NNLsRequest) returns (NNLsResponse);
  rpc mkdir(NNMkdirRequest) returns (CommonResponse);
  rpc get(NNGetRequest) returns (NNGetResponse);
  rpc put(NNPutRequest) returns (NNPutResponse);
  // datanode notice namenode blk storeed ok;
  rpc put_ok(NNPutOKRequest) returns (CommonResponse);
  rpc get_master(Request) returns (NNGetMasterResponse);
};

message DNGetRequest {
  string blkid = 1;
}
message DNGetResponse {
  Response common = 1;
  bytes data = 2;
};

message DNPutRequest {
  string blkid = 1;
  bytes data = 2;
};

message DNGetNamenodesResponse {
  Response common = 1;
  repeated string nodes = 2;
};

service DatanodeService {
  rpc put(DNPutRequest) returns (CommonResponse);
  rpc get(DNGetRequest) returns (DNGetResponse);
  rpc get_namenodes(Request) returns (DNGetNamenodesResponse);
};

service CommonService {
  rpc echo(Request) returns (CommonResponse);
}

